### PR TITLE
gcp-ilb: resource wrapping nc or socat to respond to health checks

### PR DESCRIPTION
**Pull request by Lsubatin**
This is a wrapper for nc or socat (depending on availability) to reply to health check probes. The only GCP specific piece is the optional default logging using the gcloud CLI.

### CODE REVIEW

**Comment by oalbrigt** [comment]
Isnt this basically the same as the azure-lb agent?

**Comment by Lsubatin** [comment]
The end goal is the same and it can be used for any cloud vendor that needs this kind of wrapper. 

We added error logging to an external tool (this is optional, but we use it for Stackdriver), the check for binaries is based on availability of the binary instead of the release, the handling of the pidfile is different and the socat/nc command is launched with disown (as you probably have noticed).

File: heartbeat/gcp-ilb [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+exit 0
+}
+
+#######################################################################
+
+log() {
+	lvl=$1
+	msg=$2
+	ocf_log $lvl "$0 - $process - $pid: $msg"
+	if [[ "${OCF_RESKEY_log_enable}" = "true" ]] ; then
```
Comment (oalbrigt) [inline heartbeat/gcp-ilb:158]: You should use `if ocf_is_true "${OCF_RESKEY_log_enable}"` here to be consistent with other agents supporting yes/true/1/etc.

File: heartbeat/gcp-ilb [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+	    ( ${OCF_RESKEY_log_cmd} ${OCF_RESKEY_log_params} "L $lvl: $msg" ${OCF_RESKEY_log_end_params} )
+	fi
+}
+
+getpid() {
+	if ! [[ -f "$pidfile" ]] ; then
```
Comment (oalbrigt) [inline heartbeat/gcp-ilb:164]: Can you update all the double brackets to single brackets (which is compatible with all sh and sh-based shells)?

File: heartbeat/gcp-ilb [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+	done
+
+	log err "The cat has more than 9 lives and could not be terminated."
+	return $OCF_ERR_GENERIC
+	
+}
+
+ilb_validate() {
+	check_binary "$OCF_RESKEY_cat"
+    check_binary "$OCF_RESKEY_log_cmd"
```
Comment (oalbrigt) [inline heartbeat/gcp-ilb:283]: Update to tab here to follow the indentation of the rest of the code.

Comment (Lsubatin) [reply heartbeat/gcp-ilb]: Done and committed fix. Thanks for reviewing

**Comment by Lsubatin** [comment]
Hi @oalbrigt can you (or anyone) please review and merge or comment? Thanks

**Comment by oalbrigt** [comment]
ok to test

**Comment by oalbrigt** [comment]
Can you rebase and force-push? Seems like CI is failing due to change that was done this week.

File: heartbeat/gcp-ilb [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+
+
+# Initialization:
+: ${OCF_FUNCTIONS_DIR=${OCF_ROOT}/lib/heartbeat}
+. ${OCF_FUNCTIONS_DIR}/ocf-shellfuncs
+
+# Defaults
+OCF_RESKEY_cat_default="socat"
+OCF_RESKEY_port_default="60000"
+OCF_RESKEY_log_enable="false"
```
Comment (oalbrigt) [inline heartbeat/gcp-ilb:30]: These should all be default, processed below and used as default value in the metadata section.

**Comment by Lsubatin** [comment]
Hi @oalbrigt rebasing the master in this branch replayed 94 commits into this PR. Do you want me to proceed anyway or you'd rather I submit a new PR? thanks.

**Comment by oalbrigt** [comment]
If you did `git rebase upstream/master` it should be fine. Try `git push -f` and see that it shows up here with your commits only or not.

A new PR is also fine, but it's nice to try and see how (and that) it works.

**Comment by Lsubatin** [comment]
Hi @oalbrigt can you please check again and merge? Thanks

**Comment by oalbrigt** [comment]
LGTM. Thanks.

**Pull request merged**